### ISSUE TITLE
Rule Change: Change `no-restricted-imports` to support glob pattern matching instead of gitignore-style

### ISSUE THREAD

**Issue by ItamarMalka** [eslint/eslint#16916]
### What rule do you want to change?

no-restricted-imports

### What change to do you want to make?

Implement suggestions

### How do you think the change should be implemented?

A new option

### Example code

```js
for the rule: 
"patterns": [
          {
            "group": [
              "**/functions/**",
              "!**/functions/**/common"
            ]
          }
]

The code:
import {something} from 'functions/something/common'
```

### What does the rule currently do for this code?

Currently the code doesn't pass linting since the rule uses 'ignore' for doing the pattern matching, so when it checks for the subfolder 'functions/something' the linting fails...

### What will the rule do after it's changed?

The example code should pass linting as the pattern ends with  the /common as the rule permits.

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

My current idea for solving this issue would be to implement a flag on the pattern object to state that the pattern matching should be done as glob/regex e.g `"globMatching": true` with it defaulting to false.
When the flag is toggled we can use some lightweight pattern matching lib to do use in the `isResrictedPattern` check, e.g 'minimatch' which can be used to achieve this.

**Comment by Rec0iL99** [comment]
Hi @ItamarMalka, thanks for opening the issue. 

> The example code should pass linting as the pattern ends with the /common as the rule permits.

I'm quoting the gitignore spec here:

> A leading ** followed by a slash means match in all directories. For example, `**/foo` matches file or directory `foo` anywhere, the same as pattern `foo`. `**/foo/bar` matches file or directory `bar` anywhere that is directly under directory `foo`.

The example code does not pass linting because `functions` folder matches with the `**/functions/**` pattern (according to the .gitignore spec), hence triggering the ESLint rule. 

According to the spec:

```js
// pattern: **/functions/**

// will match
import {something} from 'functions/something/common' 

// will match
import {something} from 'api/private/functions/something/common'
```

Both will trigger the ESLint rule.

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiaW1wb3J0IHtzb21ldGhpbmd9IGZyb20gJ2Z1bmN0aW9ucy9zb21ldGhpbmcvY29tbW9uJyBcblxuaW1wb3J0IHtzb21ldGhpbmcxfSBmcm9tICdhcGkvcHJpdmF0ZS9mdW5jdGlvbnMvc29tZXRoaW5nL2NvbW1vbiciLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXJlc3RyaWN0ZWQtaW1wb3J0cyI6WyJlcnJvciIseyJwYXR0ZXJucyI6W3siZ3JvdXAiOlsiKiovZnVuY3Rpb25zLyoqIl19XX1dfSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==)

**Comment by sam3k** [comment]
@ItamarMalka if I understand correctly; you want the following outcome:

🟢 Allow:`import {something} from 'functions/something/common';`
🔴 Disallowed:`import {something1} from 'functions/something/else';`
🟢 Allow:`import {something} from 'api/private/functions/something/common';`
🔴 Disallowed:`import {something1} from 'api/private/functions/something/else';`

If so, I [have not been](https://eslint.org/play/#eyJ0ZXh0IjoiaW1wb3J0IHtzb21ldGhpbmcwfSBmcm9tICdmdW5jdGlvbnMvc29tZXRoaW5nL2Vsc2UnOyAvLyBzaG91bGQgbm90IGJlIGFsbG93ZWRcblxuaW1wb3J0IHtzb21ldGhpbmd9IGZyb20gJ2Z1bmN0aW9ucy9zb21ldGhpbmcvY29tbW9uJzsgLy8gc2hvdWxkIGJlIGFsbG93ZWRcblxuaW1wb3J0IHtzb21ldGhpbmcxfSBmcm9tICdhcGkvcHJpdmF0ZS9mdW5jdGlvbnMvc29tZXRoaW5nL2NvbW1vbic7IC8vIHNob3VsZCBub3QgYmUgYWxsb3dlZFxuXG5pbXBvcnQge3NvbWV0aGluZzJ9IGZyb20gJ2FwaS9wcml2YXRlL2Z1bmN0aW9ucy9zb21ldGhpbmcvY29tbW9uJzsgLy8gc2hvdWxkIG5vdCBiZSBhbGxvd2VkXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXJlc3RyaWN0ZWQtaW1wb3J0cyI6WyJlcnJvciIseyJwYXR0ZXJucyI6W3siZ3JvdXAiOlsiZnVuY3Rpb25zLyoqLyoiLCIhZnVuY3Rpb25zLyoqL2NvbW1vbiIsIioqL2Z1bmN0aW9ucy8qKi8qIiwiISoqL2Z1bmN0aW9ucy8qKi9jb21tb24iXX1dfV19LCJlbnYiOnsiZXM2Ijp0cnVlfX19)  able to get this to work. Please let us know if this is what you are trying to accomplish.

**Comment by Rec0iL99** [comment]
@ItamarMalka I'm closing this as @sam3k was able to find a solution for us. Feel free to re-open if more issues come up. Thanks.
